internal/provisioner/common/bundle.go

```diff
@@ -0,0 +1,160 @@
+			l.Error(err, "failed to update status")
+		}
+	}()
+	u.UpdateStatus(updater.EnsureObservedGeneration(bundle.Generation))
+
+	finalizerResult, err := r.Finalizers.Finalize(ctx, bundle)
+	if err != nil {
+		u.UpdateStatus(
+			updater.EnsureResolvedSource(nil),
+			updater.EnsureContentURL(""),
+			updater.SetPhase(rukpakv1alpha1.PhaseFailing),
+			updater.EnsureCondition(metav1.Condition{
+				Type:    rukpakv1alpha1.TypeUnpacked,
+				Status:  metav1.ConditionUnknown,
+				Reason:  rukpakv1alpha1.ReasonProcessingFinalizerFailed,
+				Message: err.Error(),
+			}),
+		)
+		return ctrl.Result{}, err
+	}
+	var (
+		finalizerUpdateErrs []error
+	)
+	// Update the status subresource before updating the main object. This is
+	// necessary because, in many cases, the main object update will remove the
+	// finalizer, which will cause the core Kubernetes deletion logic to
+	// complete. Therefore, we need to make the status update prior to the main
+	// object update to ensure that the status update can be processed before
+	// a potential deletion.
+	if finalizerResult.StatusUpdated {
+		finalizerUpdateErrs = append(finalizerUpdateErrs, r.Status().Update(ctx, bundle))
+	}
+	if finalizerResult.Updated {
+		finalizerUpdateErrs = append(finalizerUpdateErrs, r.Update(ctx, bundle))
+	}
+	if finalizerResult.Updated || finalizerResult.StatusUpdated || !bundle.GetDeletionTimestamp().IsZero() {
+		err := apimacherrors.NewAggregate(finalizerUpdateErrs)
+		if err != nil {
+			u.UpdateStatus(
+				updater.EnsureResolvedSource(nil),
+				updater.EnsureContentURL(""),
+				updater.SetPhase(rukpakv1alpha1.PhaseFailing),
+				updater.EnsureCondition(metav1.Condition{
+					Type:    rukpakv1alpha1.TypeUnpacked,
+					Status:  metav1.ConditionUnknown,
+					Reason:  rukpakv1alpha1.ReasonProcessingFinalizerFailed,
+					Message: err.Error(),
+				}),
+			)
+		}
+		return ctrl.Result{}, err
+	}
+
+	unpackResult, err := r.Unpacker.Unpack(ctx, bundle)
+	if err != nil {
+		return ctrl.Result{}, UpdateStatusUnpackFailing(&u, fmt.Errorf("source bundle content: %v", err))
+	}
+
+	if r.ParseUnpackState == nil {
+		r.ParseUnpackState = defaultParseUnpackStateFunc
+	}
+
+	return r.ParseUnpackState(ctx, unpackResult, &u, bundle, r.Storage)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ProvisionerBundleReconciler) SetupWithManager(mgr ctrl.Manager, predicates ...predicate.Predicate) error {
+	l := mgr.GetLogger().WithName("controller.bundle")
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&rukpakv1alpha1.Bundle{}, builder.WithPredicates(
+			append(predicates, util.BundleProvisionerFilter(r.ProvisionerID))...,
+		)).
+		// The default unpacker creates Pod's ownerref'd to its bundle, so
+		// we need to watch pods to ensure we reconcile events coming from these
+		// pods.
+		Watches(&crsource.Kind{Type: &corev1.Pod{}}, util.MapOwneeToOwnerProvisionerHandler(context.TODO(), mgr.GetClient(), l, r.ProvisionerID, &rukpakv1alpha1.Bundle{})).
```

Comment:
The watches required by a reconciler are at least partially dependent on the unpacker implementation, which is configurable on the reconciler. With that in mind, I'd lean toward making `SetupWithManager` implementation specific.